docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,233 @@
+
+`Stubs.g.cs`:
+
+``` CSharp
+/* E */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount)
+{
+    unsafe
+    {
+        long result = 0;
+        bool success = QueryPerformanceCounter(&result) != 0;
+        lpPerformanceCount = result;
+        return success;
+    }
+}
+
+[DllImportAttribute("Kernel32.dll")]
+/* F */ private static extern int QueryPerformanceCounter(long* lpPerformanceCount);
+```
+
+The Source Generator would generate the implementation of the partial method (E) in a separate translation unit (`Stubs.g.cs`). At point (F) the `DllImportAttribute` declaration from the user's original declaration is copied into a private P/Invoke specifically for the generated code. The P/Invoke signature from the original declaration would be modified to contain only [blittable types][blittable_link] to ensure the JIT could inline the invocation. Finally note that the user's original function signature would remain in to avoid impacting existing callsites.
+
+In this system it is not defined how marshaling of specific types would be performed. The built-in runtime has complex rules for some types, and it is these rules that once shipped become the de facto standard - often times regardless if the behavior is a bug or not. The design here is not concerned with how the arguments go from a managed to unmanaged environment.
+
+### Source Generator constraint mitigations
+
+In the current Source Generator design modification of any user written code is not permitted. This includes modification of any non-functional metadata (e.g. Attributes). The above design mentions modification of user source so the current design may be untenable. Let us define the following options and describe how the above design would be modified.
+
+* In-memory source modification - This would yield the ideal design above.
+
+* On disk source modification - Altering source on disk would require the `GenerateNativeImportAttribute` to contain all properties from `DllImportAttribute`. This is required since removing the `DllImportAttribute` would be impossible and thus would cause issues during build since the method must also be marked as `partial`.
+
+* No source modification - This aligns with the current Source Generator constraints. It would be possible to create a [Roslyn Analyzer and Code fix](https://github.com/dotnet/roslyn/wiki/Getting-Started-Writing-a-Custom-Analyzer-&-Code-Fix) to aid the developer in converting their `DllImportAttribute` marked functions to use `GenerateNativeImportAttribute`. Furthermore, the function would need to be updated to have the `partial` keyword and potentially the enclosing class.
+
+## Proposed APIs
+
+The ideal solution with support for minor user defined source updates using Source Generators:
```

Comment:
I think that the ideal solution would be to just enable the source generator at the .csproj level. Adding the `GenerateNativeImportAttribute` to every DllImport is unnecessary ceremony. Why would somebody want to mark some PInvokes in the project with this, but not others?

Reply:
Off the top of my head, bugs. Having a simple way to target which functions is useful for experimentation and testing. I am not against having some property that just does it, but I would prefer to keep it a bit manual until we understand how people plan to use it.

Reply:
If it is to workaround bugs, the attribute should be opt-out (e.g. tell the generator to not touch it).

Reply:
Having it be opt-out would be dependent on in place source generate updates, right?

Reply:
Right. Any design that keeps DllImportAttribute as is would depend on source generators being able to do targeted updates of the program.

Reply:
I think this is a further potential optimization. I don't see the need to include it at present, unless the statement in the proposal of "ideal solution" is the concern.

I personally don't think that source generators going in and updating things without attributes is a good idea. I get the argument for reduced metadata, but that is easy to handle using the IL Linker. I don't like the idea of properties in a project file that magically convert things in my code. Let the code do the talking, not some build property.

Reply:
It would also be fairly trivial to extend the attribute to support `Assembly`, `Class`, or `Struct`, in which case you could annotate less while still getting the opt-in benefits.

Reply:
Oh that is interesting. I hadn't thought of that. I think that would have to be the scenario with in-memory source updates right?

Reply:
Right, I believe so, as the actual `DllImport` needs to not be the version using `out` and `bool`, etc.

Reply:
BTW: The problem with the updates can be solved by a trivial https://github.com/Fody/Fody plugin that hooks up the source-generated implementation.

I am not proposing the IL rewriters as the right solution, but I believe that community projects out there will tend to resort to trivial IL rewriters if source generators do not provide the right tools for hooking up the source-generated implementation.

Reply:
>I am not proposing the IL rewriters as the right solution, but I believe that community projects out there will tend to resort to trivial IL rewriters if source generators do not provide the right tools for hooking up the source-generated implementations.

I don't think the community should have to rely on any additional tooling other than source generators. In fact this approach would be a failure in my opinion if that isn't possible. We need to language to be able to do this. With the original proposal here this wasn't an issue. Do we think it will be possible to address with this current approach?

Reply:
> I don't think the community should have to rely on any additional tooling other than source generators. In fact this approach would be a failure in my opinion if that isn't possible. We need to language to be able to do this. With the original proposal here this wasn't an issue. Do we think it will be possible to address with this current approach?

At least given the design doc in https://github.com/dotnet/roslyn/blob/features/source-generators/docs/features/source-generators.md, the existence of tooling like `ClangSharpPInvokeGenerator`, and the current state of P/Invoke support in .NET...

I'm not sure I would have much use for this proposal in particular, but I can see where it or something similar would be useful for some people.

For my personal projects, I was just planning on migrating my current infrastructure, which uses the existing tooling to generate blittable P/Invokes, from a manual process (must manually regenerate sources) to an automated process (it uses source generators and up-to-date checks, with the generated files being deterministic and checked in).

You can see an example of where I use `ClangSharp` and an `rsp` file to generate bindings for D2D1 today here: https://github.com/terrafx/terrafx.interop.windows/blob/master/generation/um/d2d1/generate.rsp, which generates all of https://github.com/terrafx/terrafx.interop.windows/tree/master/sources/Interop/D2D1/um/d2d1. You can also see examples of much of the rest of DirectX in neighboring files, and this extends to Vulkan, X11, Wayland, and others for other repos I have.

I've customized it to take one or more header files (generally just one) as inputs and to generate a file per type, with all generated types being `partial` so you can easily extend or customize functionality where needed.

Automatically generating "safe" APIs above this is hard, because given just a C/C++ header file, it is basically impossible to know if a given `T*` is a scalar pointer, an array, an output buffer, an input buffer, etc. So that has been mostly a manual process and I would believe requires a different type of proposal/source generator to accomplish automatically.

Reply:
@tannergooding Starting from a C/C++ header is a much different issue here though. I agree that will require some additional tooling. The idea here is P/Invoke stubs being created in an AOT friendly way. The complete round trip is much hard as mentioned.

Reply:
Right, I was basically trying to say that this will likely be useful for people who already have P/Invoke signatures which are not blittable (they would have stubs generated for usages of `bool`, `char`, `ref`, `out`, `in`, and `T[]` or similar), but that it might require creative thinking to make work given the currently proposed source generator semantics.

Other people are generating bindings automatically from C/C++ headers and this likely wouldn't apply to them.

One way it could work with the existing semantics is to just have the desired public signatures (the ones taking `out` and such) in some shim-class that isn't actually used (or maybe as a `cs` input file that isn't part of the compilation?) in which case you don't need to worry about overwriting existing signatures/code.

Reply:
> I don't think the community should have to rely on any additional tooling other than source generators. In fact this approach would be a failure in my opinion if that isn't possible. 

+1

> With the original proposal here this wasn't an issue.

The original proposal required reflection at runtime. The point of the source generators is to shift reflection and other expensive stuff from runtime to compile time. If you have to use reflection at runtime to use source generators effectively, it would be also a failure.

Reply:
>The original proposal required reflection at runtime. The point of the source generators is to shift reflection and other expensive stuff from runtime to compile time. If you have to use reflection at runtime to use source generators effectively, it would be also a failure.

I agree with that in principle. In the ideal end result all of this would align and language features would exist to make the scenario work and no reflection would be needed. We would also be able to use function pointers since that seems to be the most direct method to invoke an export - please correct me here.

All that aside, I would like to understand how much latitude we want tolerate in order to reach the above ideal. My starting premise in this design proposal is stated as:

> The desire is then to provide a way that existing code can continue to function but can be modified in a way that allows it to leverage this new compiler feature.

If this isn't how we should be thinking about this, I would like to ensure we have our Goals and Non-Goals defined. I will work on updating this proposal with those so we all can agree on what is and isn't important.

Reply:
> Goals

- Allow PInvoke interop evolution independently on runtime
- High performance: No reflection at runtime, naturally AOTable

> Non-Goals 

- 100% parity with existing PInvoke marshaling rules (number of existing marshalers are used very rarely). This is what the 

> The desire is then to provide a way that existing code can continue to function but can be modified in a way that allows it to leverage this new compiler feature.

We need to meet in the middle on this: The goal for source generators should be to make this feature successful. And the goal for this feature should be to use source generators.

> We would also be able to use function pointers since that seems to be the most direct method to invoke an export - please correct me here.

I do not think the function pointers should be at the center of this feature. The most direct, the most performant and the most secure way to invoke an export is via the OS supplied import/export mechanism. This assumes AOT - Mono/Xamarin and CoreRT have this option. The current DllImport makes it possible with the right flags (`ExactSpelling=true`). If you make function pointers to be at the core part of this feature, this is not longer possible.